File: data_structure/queue/linkedlist/c/queue.c
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>

#include "queue.h"

void init(Queue *q)
{
	q->head = q->tail = NULL;
	q->size = 0;
}

void enqueue(Queue *q, elem data)
{
	Node *newNode = malloc(sizeof(*newNode));
	newNode->next = NULL;
	newNode->data = data;

	if(q->tail == NULL)
	{
		q->head = q->tail = newNode;
		return;
	}

	q->tail->next = newNode;
	q->tail = newNode;

	++(q->size);
}

void dequeue(Queue *q)
{
	if(q->head == NULL)
	{
		printf("queue is empty\n");
		assert(0);
	}
	
	Node *temp = q->head;
	q->head = q->head->next;

	if(q->head==NULL)
	{
		q->tail = NULL;
	}

	free(temp);

}

elem front(const Queue * const q)
{
	assert(q->head!=NULL);
	return q->head->data;
}

elem rear(const Queue * const q)
{
	assert(q->tail!=NULL);
	return q->tail->data;
}

void destroy(Queue *q)
{
	Node *temp = q->head;

	while(temp)
	{
		q->head = temp->next;
		free(temp);
		temp = q->head;
	}
	q->head = q->tail = NULL;
}

void display(const Queue * const q)
{
	Node *temp = q->head;

	while(temp != q->tail)
	{
		printf("%d -> ", temp->data);
		temp = temp->next;
	}

	printf("%d\n", q->tail->data);
}

File: old_ds/ruby/queue/queue.rb
# Linked list Queue Implementation
class Node
  attr_accessor :next, :data

  def initialize(data)
    @data = data
    @next = nil
  end
end

class Queue
  def initialize
    @front = nil
    @rear = nil
    @size = 0
  end

  def enqueue(data)
    new_node = Node.new(data)

    # rear |- - - - -| front
    if @size == 0
      @rear = @front = new_node
      @rear.next = @front
    else
      new_node.next = @rear
      @rear = new_node
    end

    puts "enqueue #{data}"
    @size += 1
  end

  def dequeue
    data = @front.data

    curr = @rear
    while curr.next != @front
      curr = curr.next
    end
    curr.next = nil
    @front = curr

    puts "dequeue #{data}"
    @size -= 1
  end

  def front
    @front.data
  end

  def rear
    @rear.data
  end

  def print_queue
    print "REAR--| "
    curr = @rear
    while curr != @front
      print "#{curr.data} -> "
      curr = curr.next
    end
    puts "#{@front.data} |--FRONT"
  end
end

q = Queue.new()
1.upto(10) do |x|
  q.enqueue(x)
end
q.print_queue

q.dequeue
q.dequeue
q.dequeue
q.dequeue
q.dequeue
q.print_queue

q.enqueue 100
q.print_queue

puts q.front
puts q.rear
